pkg/providers/vsphere/setupuser/setupuser.go

```diff
@@ -0,0 +1,185 @@
+	CreateRole(ctx context.Context, name string, privileges []string) error
+	SetGroupRoleOnObject(ctx context.Context, principal string, role string, object string, domain string) error
+}
+
+// SetupGOVCEnv creates appropriate govc environment variables to build govc client.
+func SetupGOVCEnv(ctx context.Context, vsuc *VSphereSetupUserConfig) error {
+	os.Setenv("GOVC_URL", vsuc.Spec.Connection.Server)
+	os.Setenv("GOVC_INSECURE", strconv.FormatBool(vsuc.Spec.Connection.Insecure))
+	os.Setenv("GOVC_DATACENTER", vsuc.Spec.Datacenter)
+	return nil
+}
+
+// Run sets up a vSphere user with appropriate group, role, and permissions to create EKS-A kubernetes clusters.
+func Run(ctx context.Context, vsuc *VSphereSetupUserConfig, govc GovcClient) error {
+	err := createGroup(ctx, vsuc, govc)
+	if err != nil {
+		return err
+	}
+
+	err = addUserToGroup(ctx, vsuc, govc)
+	if err != nil {
+		return err
+	}
+
+	err = createRoles(ctx, vsuc, govc)
+	if err != nil {
+		return err
+	}
+
+	err = associateRolesToObjects(ctx, vsuc, govc)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func createGroup(ctx context.Context, vsuc *VSphereSetupUserConfig, govc GovcClient) error {
+	exists, err := govc.GroupExists(ctx, vsuc.Spec.GroupName)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		err = govc.CreateGroup(ctx, vsuc.Spec.GroupName)
+	} else {
+		logger.V(0).Info(fmt.Sprintf("Skipping creating %s because it already exists", vsuc.Spec.GroupName))
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func createRoles(ctx context.Context, vsuc *VSphereSetupUserConfig, govc GovcClient) error {
+	for _, r := range getRoles(vsuc) {
+		exists, err := govc.RoleExists(ctx, r.name)
+		if err != nil {
+			return err
+		}
+
+		if !exists {
+			err = govc.CreateRole(ctx, r.name, r.privs)
+			if err != nil {
+				logger.V(0).Info(fmt.Sprintf("Failed to create %s role with %v", r.name, r.privs))
+				return err
+			}
+			logger.V(0).Info(fmt.Sprintf("Created %s role", r.name))
+		} else {
+			logger.V(0).Info(fmt.Sprintf("Skipping creating %s role because it already exists", r.name))
+		}
+	}
+
+	return nil
+}
+
+func associateRolesToObjects(ctx context.Context, vsuc *VSphereSetupUserConfig, govc GovcClient) error {
+	err := setGroupRoleOnObjects(ctx, vsuc, govc, vsuc.Spec.GlobalRole, []string{vSphereRootPath})
+	if err != nil {
+		return err
+	}
+
+	adminRoleObjects := append(vsuc.Spec.Objects.Folders, vsuc.Spec.Objects.Templates...)
+	err = setGroupRoleOnObjects(ctx, vsuc, govc, vsuc.Spec.AdminRole, adminRoleObjects)
+	if err != nil {
+		return err
+	}
+
+	userRoleObjects := getUserRoleObjects(vsuc)
+	err = setGroupRoleOnObjects(ctx, vsuc, govc, vsuc.Spec.UserRole, userRoleObjects)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func addUserToGroup(ctx context.Context, vsuc *VSphereSetupUserConfig, govc GovcClient) error {
+	// associate user to group
+	err := govc.AddUserToGroup(ctx, vsuc.Spec.GroupName, vsuc.Spec.Username)
```

Comment:
just like for the Role, have we tested to see that if the user is not found (if someone used the force flag accidentally with a non-existent user), err will truly be non-nil?

Reply:
Great question. Yes, error is non-nil when the user does not exist:
```
Error: govc returned error govc: user "jwmeier-22-10-17-1" not found
```